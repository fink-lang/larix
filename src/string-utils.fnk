unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
