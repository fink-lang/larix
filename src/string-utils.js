import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

