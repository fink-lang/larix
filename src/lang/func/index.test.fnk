{describe, it, eq, snapshot} = import '../../testing/jest'

{parse_expr} = import '../..'


describe:: 'func: fn ...: ...', fn:
  it:: 'parses no arg: fn: foo', fn:
    parse_expr(`fn: foo`) eq snapshot


  it:: 'parses single arg: fn foo: bar', fn:
    parse_expr(`fn foo: bar`) eq snapshot


  it:: 'parses single line: fn foo, bar: foo', fn:
    parse_expr(`fn foo, bar: foo`) eq snapshot


  it:: 'parses multi line: fn foo, bar=1: foo', fn:
    parse_expr(`
      ---
      doc comment
      ---
      fn foo, bar=1:
        foo
        bar
        # trailing func comment
    `) eq snapshot
