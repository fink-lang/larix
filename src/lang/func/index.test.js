import {parse_expr} from '../..';
import {strip_block} from '../../string-utils';


describe('func: fn ...: ...', ()=> {
  it('parses single line: fn foo, bar: foo', ()=> {
    expect(
      parse_expr(`fn foo, bar: foo`)
    ).toEqual({
      type: 'block',
      op: 'fn',
      args: [
        parse_expr('   foo'),
        parse_expr('        bar')
      ],
      exprs: [
        parse_expr('             foo')
      ],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 16, line: 1, column: 16}
      }
    });
  });


  it('parses multi line: fn foo, bar: foo', ()=> {
    expect(
      parse_expr(strip_block`
        fn foo, bar:
          foo
          bar
      `)
    ).toEqual({
      type: 'block',
      op: 'fn',
      args: [
        parse_expr('   foo'),
        parse_expr('        bar')
      ],
      exprs: [
        parse_expr('            \n  foo'),
        parse_expr('            \n     \n  bar')
      ],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 24, line: 3, column: 5}
      }
    });
  });
});
