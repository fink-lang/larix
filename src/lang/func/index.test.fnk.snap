// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`func: fn ...: ... parses multi line args 1`] = `
block fn (1:0-6:0)
  ident (2:2-2:5) foo
  assign = (3:2-3:8)
    ident (3:2-3:5) bar
    ident (3:6-3:8) ni
  :
  arithm + (5:2-5:11)
    ident (5:2-5:5) foo
    ident (5:8-5:11) bar
`;

exports[`func: fn ...: ... parses multi line: fn foo, bar=1: foo 1`] = `
--- 
--- doc comment
--- 
block fn (4:0-8:0)
  ident (4:3-4:6) foo
  assign = (4:8-4:13)
    ident (4:8-4:11) bar
    number (4:12-4:13) 1
  :
  ident (5:2-5:5) foo
  ident (6:2-6:5) bar
  # trailing func comment
`;

exports[`func: fn ...: ... parses no arg: fn: foo 1`] = `
block fn (1:0-1:7)

  :
  ident (1:4-1:7) foo
`;

exports[`func: fn ...: ... parses single arg: fn foo: bar 1`] = `
block fn (1:0-1:11)
  ident (1:3-1:6) foo
  :
  ident (1:8-1:11) bar
`;

exports[`func: fn ...: ... parses single line: fn foo, bar: foo 1`] = `
block fn (1:0-1:16)
  ident (1:3-1:6) foo
  ident (1:8-1:11) bar
  :
  ident (1:13-1:16) foo
`;
